Lint command of Composite Resource Definitions

undefinedhuman opened the pull request:
### Description of your changes

I implemented a first version of a linter for XRDs mentioned in #5393. For now I implemented three rules:

- Check for boolean fields (and suggest to change to type enum)
- Check for required fields (and suggest to make them optional with a default)
- Check fields for missing descriptions

Partially implements #5393 

Documentation issue: https://github.com/crossplane/docs/issues/917

Feedback is highly welcome. I have many more ideas for rules, but I would like to release a first version and gather feedback.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [ ] ~Added or updated unit tests.~
- [ ] ~Added or updated e2e tests.~
- [x] Linked a PR or a [docs tracking issue] to [document this change].
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~
- [ ] ~Followed the [API promotion workflow] if this PR introduces, removes, or promotes an API.~

negz commented on cmd/crank/beta/lint/cmd.go:
```diff
@@ -0,0 +1,203 @@
+
+	// Flags.
+	Output             string `default:"stdout"                      help:"Output format. Valid values are 'stdout' or 'json'. Default is 'stdout'." short:"o"`
+	SkipSuccessResults bool   `help:"Skip printing success results."`
```
What do you think about hardcoding this behavior? I personally like the Unix approach of returning 0 and printing nothing if the command works successfully.

undefinedhuman replied:
Like it - will do!

negz commented on cmd/crank/beta/lint/cmd.go:
```diff
@@ -0,0 +1,203 @@
+			return errors.Wrapf(err, "cannot convert to yaml %q", resource.Value)
+		}
+
+		var k8s unstructured.Unstructured
```
```suggestion
		var xrd unstructured.Unstructured
```

I'd prefer this name - better describes what the unstructured resource is.

negz commented on cmd/crank/beta/lint/cmd.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+func printJSON(o *output, k *kong.Context) error {
+	data, err := json.Marshal(o)
+	if err != nil {
+		return errors.Wrap(err, "cannot marshal output to JSON")
+	}
+	_, err = fmt.Fprintln(k.Stdout, string(data))
+	if err != nil {
+		return errors.Wrap(err, "cannot print summary")
+	}
+	return nil
```
Nit: I think you could use `json.NewEncoder(k.Stdout)` here. I don't feel strongly though if you prefer your current approach.

https://pkg.go.dev/encoding/json#Encoder

negz commented on cmd/crank/beta/lint/lint.go:
```diff
@@ -0,0 +1,123 @@
+	"fmt"
+	"strings"
+
+	"gopkg.in/yaml.v3"
```
Do you know if this'll also work for JSON manifests (given YAML is a superset of JSON)? JSON Kubernetes manifests are pretty rare but it'd still be nice to support them.

undefinedhuman replied:
The lib I am not sure (you can probably convert it) - I did it via JSON first (converting YAML to JSON to also check against the schema) - I wanted to have the lines though (and for the nolint rule also comments in the YAML and could not figure out how to do it properly with JSON) but yea the rule checking would also work with JSON

negz commented on cmd/crank/beta/lint/lint.go:
```diff
@@ -0,0 +1,123 @@
+				Line:    key.Line,
+				Error:   false,
+				RuleID:  "XRD001",
+				Message: fmt.Sprintf("Boolean field detected at path %s — consider using an enum instead for extensibility.", path),
```
What do you think about the Issue structure containing an optional link to more info? I could imagine linking to the upstream Kube API conventions and/or a crossplane.io docs page for this linter rule.

undefinedhuman replied:
Love it - will add it

undefinedhuman replied:
I started with the general concept and implemented a reference to the CRD design spec for the boolean fields - as soon as I written the docs I can also add the links to the other rules

negz commented on cmd/crank/beta/lint/lint.go:
```diff
@@ -0,0 +1,123 @@
+					Name:    name,
+					Line:    item.Line,
+					RuleID:  "XRD002",
+					Message: fmt.Sprintf("Required field '%s' at path %s.%s — consider making it optional with a default.", item.Value, path, key.Value),
```
Ideally we'd support a `# nolint` directive, e.g. for required fields that really should be required. That could be a future improvement though.

undefinedhuman replied:
Already on the list!

undefinedhuman replied:
I implemented a "simple check" a comment above the nodes (checking for either `# nolint` or `# nolint RULEID`) - please check the attached "lint/testdata/xrd-comments.yaml"

negz reviewed (commented):
Thanks @undefinedhuman - this is a great contribution.

undefinedhuman commented on cmd/crank/beta/lint/cmd.go:
```diff
@@ -0,0 +1,203 @@
+// Cmd arguments and flags for render subcommand.
+type Cmd struct {
+	// Arguments.
+	Resources string `arg:"" help:"Resources source which can be a file, directory, or '-' for standard input."`
```
For now the linter only works for XRDs definitely makes sense to put it in the description - other resources become increasingly complex due to the number of ways to define a composition for example - but imo definitely makes sense to also have this sort of linter for other resources

undefinedhuman replied:
I will update the help message

undefinedhuman commented:
@negz thanks for the review! It would be great if you could review the new changes again and let me know what you think. 

I have implemented your suggestions

- Changed the success output to UNIX style
- Implemented the “nolint” functionality
- Changed the JSON output to use the decoder
- Changed the help text to indicate which resources can be linked.

Next steps: I think the logical next step is the documentation, I will sit down and write it as soon as I have time (probably in the next days). I ran earthly again and fixed all new issues. For the pipelines, I'm not 100% sure, but it also looks like tests that have nothing to do with my changes?

negz commented:
Someone just shared this with me: https://github.com/kubernetes-sigs/kube-api-linter

I wonder whether we should fork and extend it for Crossplane? Their documentation for adding new linters is good, but it appears to be something you do at build time.

undefinedhuman commented:
I have no strong opinion about it yet - I did not work with the tool yet/implemented a linter for it - one benefit I see with the current implementation that it would all be a single CLI which is quite convenient for CI setups etc. using one single binary/docker image for everything crossplane related.

One thing from quickly scanning kube-api-linter it seems to be only for APIs - if we would like to expand the linter to cover compositions as well this could become problematic.

negz commented:
@undefinedhuman Agreed. I was thinking something along the lines of importing the linter and using it as a library to build our own tool. I don't feel super strongly though - just trying to avoid reinventing the wheel.

undefinedhuman commented:
I am super busy at work currently - but I can investigate the tool :)

negz commented:
@undefinedhuman I got a quick chance to chat to the kube-api-linter authors at Kubecon earlier this week. It sounds like they're open to collaborating - iirc they suggested we open an issue on their repo with our use case. Then we can discuss what changes we could make to make it easier to consume kube-api-linter.

negz commented:
CC @JoelSpeed

undefinedhuman commented:
However thinking about the implementation, imo we would need to build a setup to convert the Crossplane XRD Yamls to the respective go structs to be able to analyze them with "kube-api-linter" - do you see this as reliably feasible? Would love your input.
